cpp/src/arrow/util/thread_pool_benchmark.cc

```diff
@@ -136,21 +168,24 @@ static void ThreadedTaskGroup(benchmark::State& state) {
 
   for (auto _ : state) {
     auto task_group = TaskGroup::MakeThreaded(pool.get());
-    for (int32_t i = 0; i < nspawns; ++i) {
-      // Pass the task by reference to avoid copying it around
-      task_group->Append(std::ref(task));
-    }
+    task_group->Append([&task, nspawns, task_group] {
```

Comment:
Why the nested append?

Reply:
hold over, will clean

Reply:
It's not exactly a holdover.  `SerialTaskGroup::AppendReal` has `DCHECK(!finished_);` which makes sense (you don't want to add a task to a finished task group).  However, if you add the same `DCHECK(!finished_);` to `ThreadedTaskGroup` then the old non-nested approach in `TestTaskGroupErrors` (in task_group_test.cc) fails.

What happens is that the first one or two tasks gets added to the task group and, if they are fast, they finish before the task adder gets a chance to add the rest of the tasks and the task group gets marked finished.  This doesn't happen with the serial task group since that task group cannot be marked finished until the caller waits on this.

Ben and I had some discussion about possibly adding a `Start` method to task group but I don't remember where that ended up.  So as it is right now you need to do this nested approach to ensure that the task group does not finish until you have scheduled all the tasks you want to run.

Also, I put a commented out `DCHECK(!finished_);` in `ThreadedTaskGroup`.  I don't want to uncomment it because there are still some places using threaded task group that don't use it in the nested fashion and I don't want to break them.

Reply:
I spoke with @bkietz a bit more on this and he pointed out it is not a race condition as written without the futures.  The threaded task group cannot actually finish unless `Finish` is called.  Once futures are added in then it becomes an issue.  As soon as the task counter hits 0 it will complete the future.

I spoke about whether we should pull this nesting back out and it was recommended to just leave it all alone since futures will replace task group someday.